pandas/core/interval.py

```diff
@@ -0,0 +1,549 @@
+                                 fill_value=fill_value)
+
+        try:
+            new_left = taker(left)
+            new_right = taker(right)
+        except ValueError:
+
+            # we need to coerce; migth have NA's in an
+            # integer dtype
+            new_left = taker(left.astype(float))
+            new_right = taker(right.astype(float))
+
+        return self._shallow_copy(new_left, new_right)
+
+    take_nd = take
+
+    def _format_data(self):
+
+        # TODO: integrate with categorical and make generic
+        # name argument is unused here; just for compat with base / categorical
+        n = len(self)
+        max_seq_items = min((get_option(
+            'display.max_seq_items') or n) // 10, 10)
+
+        formatter = str
+
+        if n == 0:
+            summary = '[]'
+        elif n == 1:
+            first = formatter(self[0])
+            summary = '[{first}]'.format(first=first)
+        elif n == 2:
+            first = formatter(self[0])
+            last = formatter(self[-1])
+            summary = '[{first}, {last}]'.format(first=first, last=last)
+        else:
+
+            if n > max_seq_items:
+                n = min(max_seq_items // 2, 10)
+                head = [formatter(x) for x in self[:n]]
+                tail = [formatter(x) for x in self[-n:]]
+                summary = '[{head} ... {tail}]'.format(
+                    head=', '.join(head), tail=', '.join(tail))
+            else:
+                head = []
+                tail = [formatter(x) for x in self]
+                summary = '[{tail}]'.format(tail=', '.join(tail))
+
+        return summary
+
+    def _format_space(self):
+        space = ' ' * (len(self.__class__.__name__) + 1)
+        return "\n{space}".format(space=space)
+
+    @property
+    def left(self):
+        """
+        Return the left endpoints of each Interval in the IntervalIndex as
+        an Index
+        """
+        return self._left
+
+    @property
+    def right(self):
+        """
+        Return the right endpoints of each Interval in the IntervalIndex as
+        an Index
+        """
+        return self._right
+
+    @property
+    def closed(self):
+        """
+        Whether the intervals are closed on the left-side, right-side, both or
+        neither
+        """
+        return self._closed
+
+    @property
+    def length(self):
+        """
+        Return an Index with entries denoting the length of each Interval in
+        the IntervalIndex
+        """
+        try:
+            return self.right - self.left
+        except TypeError:
+            # length not defined for some types, e.g. string
+            msg = ('IntervalIndex contains Intervals without defined length, '
+                   'e.g. Intervals with string endpoints')
+            raise TypeError(msg)
+
+    def __repr__(self):
+        return "{}({})".format(self.__class__.__name__, self._format_data())
+
+    @property
+    def values(self):
```

Comment:
Should this be called `values` on the Array? 
I think there are in general (although already not fully applicable in the Interval case, but it is for periods) two possible arrays: the actual storage array (eg the integers in case of Periods, the record array in case of ipaddress) or a numpy array with boxed objects. 
Both should be accessible I think (although the first is maybe an specific Array class implementation detail and should only be used inside the class itself?), but not sure the second should be `.values`. 
The question is also what `__array__` should be if we add it?